curvefs/src/metaserver/metastore.cpp

```diff
@@ -146,93 +67,34 @@ bool MetaStoreImpl::Load(const std::string& pathname) {
-    auto cntr = partition->GetInodeContainer();
-    auto container = std::shared_ptr<InodeContainerType>(
-        cntr, [](InodeContainerType*) {});  // don't release storage
-    auto iterator = std::make_shared<InodeIteratorType>(ENTRY_TYPE::INODE,
-                                                        partitionId, container);
-    return iterator;
-}
-
-std::shared_ptr<Iterator> MetaStoreImpl::NewDentryIterator(
-    std::shared_ptr<Partition> partition) {
-    auto partitionId = partition->GetPartitionId();
-    auto cntr = partition->GetDentryContainer();
-    auto container = std::shared_ptr<DentryContainerType>(
-        cntr, [](DentryContainerType*) {});  // don't release storage
-    auto iterator = std::make_shared<DentryIteratorType>(
-        ENTRY_TYPE::DENTRY, partitionId, container);
-    return iterator;
-}
-
-std::shared_ptr<Iterator> MetaStoreImpl::NewPendingTxIterator(
-    std::shared_ptr<Partition> partition) {
-    PrepareRenameTxRequest pendingTx;
-    auto container = std::make_shared<PendingTxContainerType>();
-    if (partition->FindPendingTx(&pendingTx)) {
-        container->emplace(0, pendingTx);
-    }
-
-    auto partitionId = partition->GetPartitionId();
-    auto iterator = std::make_shared<PendingTxIteratorType>(
-        ENTRY_TYPE::PENDING_TX, partitionId, container);
-    return iterator;
-}
-
-void MetaStoreImpl::SaveBackground(const std::string& path,
+void MetaStoreImpl::SaveBackground(MetaStoreFStream* fstream,
+                                   const std::string& path,
+                                   std::shared_ptr<MergeIterator> iterator,
                                    OnSnapshotSaveDoneClosure* done) {
     LOG(INFO) << "Save metadata to file background.";
-
-    std::vector<std::shared_ptr<Iterator>> children;
-    auto iterator = NewPartitionIterator();  // partition
-    children.push_back(iterator);
-
-    for (const auto& item : partitionMap_) {
-        auto partition = item.second;
-
-        iterator = NewInodeIterator(partition);  // inode
-        children.push_back(iterator);
-
-        iterator = NewDentryIterator(partition);  // dentry
-        children.push_back(iterator);
-
-        iterator = NewPendingTxIterator(partition);  // pending tx
-        children.push_back(iterator);
-    }
-
-    auto mergeIterator = std::make_shared<MergeIterator>(children);
-    bool succ = SaveToFile(path, mergeIterator);
+    bool succ = fstream->Save(path, iterator);
     LOG(INFO) << "Save metadata to file " << (succ ? "success" : "fail");
+
     if (succ) {
         done->SetSuccess();
     } else {
         done->SetError(MetaStatusCode::SAVE_META_FAIL);
     }
-
     done->Run();
 }
 
 bool MetaStoreImpl::Save(const std::string& path,
                          OnSnapshotSaveDoneClosure* done) {
-    ReadLockGuard readLockGuard(rwLock_);
-    std::thread th =
-        std::thread(&MetaStoreImpl::SaveBackground, this, path, done);
+    WriteLockGuard writeLockGuard(rwLock_);
```

Comment:
I think this lock doesn't help, even you build iterators within the lock, because this lock is released after this function is returned.

And the dump works without lock, so data that iterators point to can be modified or deleted.

Reply:
yep, rocksdb's problem has been fixed by snapshot, https://github.com/opencurve/curve/issues/1228

Reply:
now for memory storage, we will hold write lock until child process forked, it can gurantee all data consistent.

Reply:
> if metastore has stored lots of data, this will block other requests, anyway, we can leave it later.

actullay, we only return the iterator of all data execpt the partition iterator which is small enough, and iterate all data in the child process, so it will not lead to a performance problem.